### PR TITLE
Revise SPF requirement policies of EXO baselines and update Rego code for the 2 updated policies

### LINKED ISSUE #900: Revise SPF Requirements
# 💡 Summary #
- Delete MS.EXO.2.1v1
- Revise MS.EXO.2.2v1 to “An SPF policy SHALL be published for each domain that fails all non-approved senders.”

## Motivation and context ##
- Maintaining a list is not configuration item (more of a prerequisite)
- There was some confusion about what to do with the list once you have it.
- Also some confusion about IP vs domains (e.g., `include:spf.protection.outlook.com`).
- "only these" is ambiguous
- Explicitly spell out that non-approved senders should fail

## Implementation notes ##
See https://github.com/cisagov/ScubaGear/issues/857 for a related technical issue.

## Acceptance criteria ##
- [ ] The baseline is revised
- [ ] The Rego is updated to reflect the change

### PR DESCRIPTION
**Pull request by nanda-katikaneni**
#  #
  

## 🗣 Description ##

Update and revise MS.EXO.2.1v1 and MS.EXO.2.2v1 baseline policies. Revisions include: deletion of MS.EXO.2.1v1 in the markdown, renaming and updating the language of MS.EXO.2.2v1 in the markdown, updating the rego code to be consistent with markdown changes, updating unit tests and functional tests to align with rego code changes.   

## 💭 Motivation and context ##
closes #900

## 🧪 Testing ##

- Verify that EXO markdown baselines are updated to delete MS.EXO.2.1v1, rename MS.EXO.2.1v1.
- Execute EXO baseline on a tenant and confirm that control group 2 has a single test and that it is showing correct result
- Execute EXO unit tests and ensure that all tests are passing
- Execute EXO functional tests and ensure that all tests are passing

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] PR targets the correct parent branch (e.g., main or release-name) for merge.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] Changes are sized such that they do not touch excessive number of files.
- [x] *All* future TODOs are captured in issues, which are referenced in code comments.
- [x] These code changes follow the ScubaGear [content style guide](https://github.com/cisagov/ScubaGear/blob/main/CONTENTSTYLEGUIDE.md).
- [x] Related issues these changes resolve are linked preferably via [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).
- [x] All relevant type-of-change labels added.
- [x] All relevant project fields are set.
- [x] All relevant repo and/or project documentation updated to reflect these changes.
- [x] Unit tests added/updated to cover PowerShell and Rego changes.
- [x] Functional tests added/updated to cover PowerShell and Rego changes.
- [x] All relevant functional tests passed.
- [ ] All automated checks (e.g., linting, static analysis, unit/smoke tests) passed.

## ✅ Pre-merge checklist ##

- [x] PR passed smoke test check.
- [x] Feature branch has been rebased against changes from parent branch, as needed

  Use `Rebase branch` button below or use [this](https://www.digitalocean.com/community/tutorials/how-to-rebase-and-update-a-pull-request) reference to rebase from the command line.
- [x] Resolved all merge conflicts on branch
- [x] Notified merge coordinator that PR is ready for merge via comment mention

## ✅ Post-merge checklist ##

- [ ] Feature branch deleted after merge to clean up repository.
- [ ] Verified that all checks pass on parent branch (e.g., main or release-name) after merge.

### CODE REVIEW

File: PowerShell/ScubaGear/baselines/exo.md [hunk 1]
```diff
@@ -113,20 +113,13 @@ points.
 ### Policies
 
 #### MS.EXO.2.1v1
-A list of approved IP addresses for sending mail SHALL be maintained.
+An SPF policy SHALL be published for each domain that fails all non-approved senders.
```
Comment (ahuynhMITRE) [inline PowerShell/ScubaGear/baselines/exo.md:116]: description of the change notes that MS.EXO.2.1v1 will be deleted and changes will be applied to MS.EXO.2.2v1. 

If this is still the case I would remove all references to MS.EXO.2.1v1 and bump MS.EXO.2.2v1 to MS.EXO.2.2v2 with the updated policy, last modified date, and instructions.

Comment (ahuynhMITRE) [reply PowerShell/ScubaGear/baselines/exo.md:116]: the subsequent changes to reference to the new MS.EXO.2.2v2 will also need to be reflected in the exoconfig.rego, exoconfig_02_test.rego, and exo.testplan.yaml files

Comment (adhilto) [reply PowerShell/ScubaGear/baselines/exo.md:116]: Relatedly, "An SPF policy SHALL be published..." appears now as MS.EXO.2.1.v1 in the markdown. Shouldn't it still be MS.EXO.2.2, but v2?

**Comment by mitchelbaker-cisa** [comment]
Other than the duplicate EXO.14.1 policy this is the last baseline related PR to merge before we submit the consolidated change document for CISA review. We'd like to get the baselines submitted within the next couple days, by end of this week at the latest. 

@nanda-katikaneni What final considerations are left for the SPF revisions?

**Comment by adhilto** [comment]
> Other than the duplicate EXO.14.1 policy this is the last baseline related PR to merge before we submit the consolidated change document for CISA review. We'd like to get the baselines submitted within the next couple days, by end of this week at the latest.
> 
> @nanda-katikaneni What final considerations are left for the SPF revisions?

@mitchelbaker-cisa 
Two additional things Nanda and I discussed offline:
1. Add this note that is in the Gmail baseline, to clarify what we mean by the ambiguous instruction to "fail." 
image

2. Add a paragraph like this to the top of the implementation: 
image
Screenshot is again from the Gmail baseline. I also think would be good to include here, as there's been some confusion about whether you include IP addresses or domain names in the SPF policy (turns out it's kind of both, as explained in the paragraph). We'd need to substitute the Gmail links with M365 documentation, but I do think it includes valuable context we'd want in EXO baseline too.

**Comment by nanda-katikaneni** [comment]
@mitchelbaker-cisa 
I should be able to checkin my updates (to address comments from @adhilto and @ahuynhMITRE) and get them re-reviewed for merge in next couple days.

File: PowerShell/ScubaGear/baselines/exo.md (hunk 1 above)
Comment (nanda-katikaneni) [reply PowerShell/ScubaGear/baselines/exo.md:121]: Modified MS.EXO.2.2v2 (deleted MS>EXO.2.1v1)

File: PowerShell/ScubaGear/baselines/exo.md [hunk 2]
```diff
@@ -158,12 +152,9 @@ agency, facilitating phishing attacks. Publishing an SPF policy for each agency
 
 ### Implementation
 
-#### MS.EXO.2.1v1 Instructions
-Identify any approved senders specific to your agency.
-Additionally, see [External DNS records required for SPF](https://learn.microsoft.com/en-us/microsoft-365/enterprise/external-domain-name-system-records?view=o365-worldwide#external-dns-records-required-for-spf) for
-inclusions required for Microsoft to send email on behalf of your domain.
+#### MS.EXO.2.2v2 Instructions
+First, identify any approved senders specific to your agency (see  [Create safe sender lists in EOP](https://learn.microsoft.com/en-us/defender-office-365/create-safe-sender-lists-in-office-365) for tips). SPF allows you to indicate approved senders by IP address or CIDR range. However, note that SPF allows you to [include](https://www.rfc-editor.org/rfc/rfc7208#section-5.2) the IP addresses indicated by a separate SPF policy, referred to by domain name. See [External DNS records required for SPF](https://learn.microsoft.com/en-us/microsoft-365/enterprise/external-domain-name-system-records?view=o365-worldwide#external-dns-records-required-for-spf) for inclusions required for M365 to send email on behalf of your domain.
```
Comment (adhilto) [inline PowerShell/ScubaGear/baselines/exo.md:156]: I don't think the safe sender lists link is relevant here, as that is a different feature entirely (safe sender list = who can my users receive email from vs SPF = who is approved to send email for _my_ domain). Honestly, we could probably get away without having a link for that first part, and it's going to vary a lot by agency. We could also just say something like, "First, identify any approved senders specific to your agency, e.g., any on-premises mail servers."

File: PowerShell/ScubaGear/Rego/EXOConfig.rego [hunk 3]
```diff
@@ -51,35 +50,22 @@ tests contains {
...
+# MS.EXO.2.2v2
 #--
 
 # Loop through domain DNS responses & loop through the policies associated
 # with the domain. Save the records that start with string "v=spf1". If
 # records does not exist, save domain name in DomainsWithoutSpf array.
 DomainsWithoutSpf contains DNSResponse.domain if {
     some DNSResponse in input.spf_records
-    SpfRecords := {Record | some Record in DNSResponse.rdata; startswith(Record, "v=spf1 ")}
+    #SpfRecords := {Record | some Record in DNSResponse.rdata; startswith(Record, "v=spf1 ")}
```
Comment (adhilto) [inline PowerShell/ScubaGear/Rego/EXOConfig.rego:61]: Recommend removing the dead code

```suggestion
```

File: PowerShell/ScubaGear/Testing/Unit/Rego/EXO/EXOConfig_02_test.rego [hunk 4]
```diff
@@ -142,14 +128,30 @@ test_Rdata_Multiple_Correct_V2 if {
...
         ]
     }
 
-    TestResult("MS.EXO.2.2v1", Output, PASS, true) == true
+    TestResult("MS.EXO.2.2v2", Output, PASS, true) == true
+}
+
+test_Rdata_Multiple_Correct_V3 if {
+    # Test SPF redirect
+    PolicyId := "MS.EXO.2.2v2"
```
Comment (adhilto) [inline PowerShell/ScubaGear/Testing/Unit/Rego/EXO/EXOConfig_02_test.rego:143]: This variable is never used and is causing the unit tests to fail.

**Review by adhilto — CHANGES_REQUESTED** [review]

**Comment by nanda-katikaneni** [comment]
@adhilto and @ahuynhMITRE, all the comments/changes are addressed - please re-review.

**Review by ahuynhMITRE — APPROVED** [review]
no additional comments from me

**Review by adhilto — APPROVED** [review]

**Pull request merged**